### PR TITLE
Improve choosing dividing-axis

### PR DESCRIPTION
**Pull request by pabloferz**
Fixes https://github.com/stevengj/HCubature.jl/issues/4

When the fourth difference is around the same value along every axis, we should choose the axis that corresponds to the largest width to ensure we sample better the whole box.

C.c. @stevengj

### CODE REVIEW

**Comment by giordano** [comment]
Maybe add a test? 😃

File: src/genz-malik.jl [hunk 1]
```diff
@@ -138,6 +138,8 @@ function (g::GenzMalik{n,T})(f, a::SVector{n}, b::SVector{n}, norm=vecnorm) wher
         if divdiff > maxdivdiff
```
Comment (pabloferz) [inline src/genz-malik.jl]: We could add a `divdiff = divdiff < eps(T) ? zero(divdiff) : divdiff` above this to ignore negligible differences and choose from the half-widths instead, but I'm not sure if its worth the trouble.

**Comment by stevengj** [comment]
The thing that bugs me about this is the precision of the test.   I feel like it should somehow be tied to an error estimate computed from `f`, but I'm not sure how to do this in a reasonable way.

**Comment by pabloferz** [comment]
Right now I'm not sure what would be an appropriate way of doing that. Does `hcubature(f, (0.0,0.0,-0.2), (0.2,2π,0.2), rtol=1e-6)[1] ≈ (22502//140625)*π rtol=1e-6` seems better than what I currently have?

**Comment by stevengj** [comment]
It certainly seems to help this particular integral, but any fix that depends on exact equality seems very fragile.

One thing we could do would be to use the current error estimate `E` in the integral as a measure of the uncertainty in `f`, i.e. `δf = E/volume`, and the do all comparisons with that absolute tolerance.

**Comment by pabloferz** [comment]
> It certainly seems to help this particular integral, but any fix that depends on exact equality seems very fragile.

Agreed.

> One thing we could do would be to use the current error estimate E in the integral as a measure of the uncertainty in f, i.e. δf = E/volume, and the do all comparisons with that absolute tolerance.

Do you mean that we could choose by width along an axis when `abs(maxdiff - divdiff) < δf`? That is somewhat along the lines that what I had proposed above (setting `divdiff = divdiff < eps(T) ? zero(divdiff) : divdiff`) except that the tolerance is independent of the box.

**Comment by pabloferz** [comment]
I pushed some changes that should make this more robust. The scheme would be the following:

1. Set some tolerance `ε`.
2. If the fourth difference `divdiff` is greater than the maximum fourth difference found so far `maxdivdiff` by at least `ε`, set the axis to the current dimension.
3. If `abs(divdiff - maxdivdiff) < ε` then subdivide along the the dimension with larger width.

I have set `ε` to the machine epsilon and not to a value depending on the error of the integral over the current box, as we would have to store the fourth differences to choose the axis at the end. I believe this should be good enough but if there's a stronger argument for setting `ε` to another value, this could be changed.

**Comment by stevengj** [comment]
I was thinking of your current PR except with `ε` replaced by `δf`.

(A tolerance independent of what the integrand is doing doesn't make sense to me.  `eps(typeof(maxdivdiff))` is an especially poor choice because it is scaled incorrectly; equivalently, it is dimensionless while `δ` is dimensionful, so your units are wrong.)

**Comment by stevengj** [comment]
Note that my suggested `δf` has the nice property (for sufficiently smooth integrands) that it goes to zero asymptotically faster than `divdiff` as the box size shrinks, so for a sufficiently small box size it is equivalent to an equality test.

**Comment by pabloferz** [comment]
Forgot about dimensions. As I commented above I set the tolerance to a fixed value to avoid storing the fourth differences and choosing the axis at the end, but that is probably not of significant cost compared with the rest of the routine.

**Comment by stevengj** [comment]
You could use`eps(norm(f₁))` to have the correct units.  However, as the integrand becomes more oscillatory, the divided differences become less accurate, which is why I suggested a `δf` that depends on how oscillatory the function is (as measured by the error estimate).

I agree that this makes the implementation more complicated, but I think it is worth it.

**Comment by pabloferz** [comment]
`δf = E / V` seems to require more function evaluations (see the failling tests) for many integrands. I tested locally using either `eps(norm(I) / V)` and `eps(E  / V)` and none of those seem to cause any problems.

**Comment by stevengj** [comment]
We can always use `δf = 0.01 * E / V` or any similar scale factor; the scaling here seems pretty heuristic.

**Comment by stevengj** [comment]
(The failed test seems to be only a 10% increase in function evaluations, though, which isn't too bad.)

**Comment by pabloferz** [comment]
I did some experimenting and found that `δf = 0.001 * E / V` gives the same function evaluations in the test example while solving the problem with the problematic example from #4.

**Comment by stevengj** [comment]
Sorry for the long delay; this looks ready to merge.

**Pull request merged**